### DISCUSSION TITLE
How does one select 10 premises in the environment during training time?

### DISCUSSION THREAD

**Discussion by brando90** [princeton-vl/CoqGym#29] [Q&A] [answered]
I saw in the paper that you choose 10 premises in the environment:

> Encoding the goal and premises ASTactic encodes the current goal and premises into vectors. We include the entire local context and up to 10 premises in the environment, ex- cluding a large number of premises imported from libraries. A model could be more powerful if it is capable of selecting relevant premises from the entire environment, but that is left for future research.

I was wondering how during training you got those 10 premises. Do you pre-select them before training and that is how the data set was created so the a data loader can get a supervised learning data point `(x=<local context, goal, env>, y=tac)` ? Or did you do something else?

I guess I am unsure how you are getting the 10 premises from a static data set without the actual proving going with Coq during training (plus to select environment terms don't you need a premise selection model or something already to be able to choose them or how are you selecting the 10 premises for training/generating the data set)?

Thanks in advance Kaiyu!

**Comment by yangky11** [comment, marked as answer]
As shown here:
https://github.com/princeton-vl/CoqGym/blob/37f93ce320eb6231758024caff39a0f6faaf6825/ASTactic/agent.py#L34-L41
, we simply take the last 10 environment terms. This is motivated by the fact that terms defined close to the proof are more likely to be useful. However, this is indeed a very poor approximation. We did this because considering all environment terms is computationally infeasible.

**Reply by brando90** [reply]
So to make sure I understand, is that (last 10 premise selection) done when you create the data set (as a preprocessing step), right?

**Reply by yangky11** [reply]
It is called when extracting proof steps from the proofs.
https://github.com/princeton-vl/CoqGym/blob/37f93ce320eb6231758024caff39a0f6faaf6825/ASTactic/extract_proof_steps.py#L50

So the CoqGym dataset contains all environment terms, but we keep only the last 10 when preprocessing it into proof steps.
